synapse/static/client/login/js/login.js

```diff
@@ -46,23 +48,36 @@ var setFeedbackString = function(text) {
 };
 
 var show_login = function() {
-    $("#loading").hide();
-
     var this_page = window.location.origin + window.location.pathname;
     $("#sso_redirect_url").val(this_page);
 
-    if (matrixLogin.serverAcceptsPassword) {
-        $("#password_flow").show();
+    if (matrixLogin.serverAcceptsSso) {
+        if (try_token() || matrixLogin.serverAcceptsPassword) {
+            // Show the SSO form if there's a login token in the query. That's because,
+            // if there is a token, and this function is run, it means an error happened,
+            // and in this case it's nicer to show the form with an error rather than
+            // redirect immediately to the SSO portal.
+            // Also show the form if the server is accepting password login as well.
+            $("#sso_form").show();
+        } else {
+            // Submit the SSO form instead of displaying it. The reason behind this
+            // behaviour is that the user will likely arrive here after clicking on a
+            // button, in the client, with a label such as "Continue with SSO", so
+            // clicking on another button with the same semantics is a bit pointless.
+            $("#sso_form").submit();
+            return;
+        }
     }
 
-    if (matrixLogin.serverAcceptsSso) {
-        $("#sso_flow").show();
-    } else if (matrixLogin.serverAcceptsCas) {
-        $("#sso_form").attr("action", "/_matrix/client/r0/login/cas/redirect");
-        $("#sso_flow").show();
+    set_title(title_pre_auth);
+
+    $("#loading").hide();
```

Comment:
stick these at the end so that the logic to handle the login types isn't interrupted by unrelated code?